### ISSUE TITLE
Implement execution mode

### ISSUE THREAD

**Issue by edho08** [edho08/synamont#55]
The engine currently only allows parallel mode of execution. instead choosing parallelism should be deferred to user. Not only that, the granularity of parallel execution is also left to the user. Parallel execution severely degrade performance in simple task hence why it should instead be optional. 

Granularity of execution should be as defined below:

1. Sequential. Completely sequential execution.
2. Handler. Each handler in an event is in parallel execution but event execution is sequential.
3. Event. Each event execution is parallel but each handler in an event execution is sequential.
4. Complete. All event and handler execution is parallel.

The internal implementation use rwlock for grained locking. in sequential execution this degrade performance a little. we cannot do anything about this
